File: src/geometry_triangle_properties_medium.py
"""
geometry - triangle_properties (medium)
Generated: 2026-02-22T04:05:06.279038
"""

from problem_utils import *
from svg_utils import Diagram

def generate():
    problem_type = randint(1, 5)
    
    if problem_type == 1:
        # Pythagorean theorem - find missing side (1300-1400)
        leg1 = randint(3, 12)
        leg2 = randint(3, 12)
        hypotenuse = sqrt(leg1**2 + leg2**2)
        
        # Make sure it's either a Pythagorean triple or a simple radical
        if hypotenuse == int(hypotenuse):
            hypotenuse = int(hypotenuse)
            
        d = Diagram(width=300, height=250, padding=40)
        d.polygon([(0, 0), (leg1, 0), (leg1, leg2)], labels=['A', 'B', 'C'])
        d.right_angle((leg1, 0), (0, 0), (leg1, leg2))
        d.segment_label((0, 0), (leg1, 0), f'{leg1}')
        d.segment_label((leg1, 0), (leg1, leg2), f'{leg2}')
        d.segment_label((0, 0), (leg1, leg2), 'c')
        
        return problem(
            question=f"In right triangle ABC, the legs have lengths {leg1} and {leg2}. Find the length of the hypotenuse $c$.",
            answer=hypotenuse,
            difficulty=(1300, 1400),
            topic="geometry/triangle_properties",
            solution=steps(
                f"Use the Pythagorean theorem: $c^2 = a^2 + b^2$",
                f"$c^2 = {leg1}^2 + {leg2}^2 = {leg1**2} + {leg2**2} = {leg1**2 + leg2**2}$",
                f"$c = \\sqrt{{{leg1**2 + leg2**2}}} = {latex(hypotenuse)}$"
            ),
            image=d.render()
        )
    
    elif problem_type == 2:
        # Triangle angle sum - find missing angle (1200-1300)
        angle1 = randint(30, 80)
        angle2 = randint(30, 90)
        while angle1 + angle2 >= 180:
            angle2 = randint(30, 90)
        angle3 = 180 - angle1 - angle2
        
        return problem(
            question=f"In triangle ABC, angle A measures ${angle1}°$ and angle B measures ${angle2}°$. Find the measure of angle C in degrees.",
            answer=angle3,
            difficulty=(1200, 1300),
            topic="geometry/triangle_properties",
            solution=steps(
                f"The sum of angles in a triangle is $180°$",
                f"$\\angle C = 180° - \\angle A - \\angle B$",
                f"$\\angle C = 180° - {angle1}° - {angle2}° = {angle3}°$"
            )
        )
    
    elif problem_type == 3:
        # Isosceles triangle - find angles (1300-1400)
        vertex_angle = randint(20, 80) * 2  # Make it even for clean division
        base_angle = (180 - vertex_angle) // 2
        
        d = Diagram(width=300, height=250, padding=40)
        height = 8
        base = 6
        d.polygon([(0, 0), (base, 0), (base/2, height)], labels=['B', 'C', 'A'])
        d.tick_marks((0, 0), (base/2, height), count=1)
        d.tick_marks((base, 0), (base/2, height), count=1)
        d.segment_label((0, 0), (base/2, height), '')
        d.segment_label((base, 0), (base/2, height), '')
        
        return problem(
            question=f"In isosceles triangle ABC with $AB = AC$, the vertex angle at A measures ${vertex_angle}°$. Find the measure of angle B in degrees.",
            answer=base_angle,
            difficulty=(1300, 1400),
            topic="geometry/triangle_properties",
            solution=steps(
                f"In an isosceles triangle, the base angles are equal",
                f"Let $\\angle B = \\angle C = x$",
                f"$\\angle A + \\angle B + \\angle C = 180°$",
                f"${vertex_angle}° + x + x = 180°$",
                f"$2x = {180 - vertex_angle}°$",
                f"$x = {base_angle}°$"
            ),
            image=d.render()
        )
    
    elif problem_type == 4:
        # Triangle perimeter given sides (1200-1300)
        side1 = randint(4, 12)
        side2 = randint(4, 12)
        side3 = randint(max(abs(side1 - side2) + 1, 4), side1 + side2 - 1)
        perimeter = side1 + side2 + side3
        
        return problem(
            question=f"A triangle has sides of length ${side1}$, ${side2}$, and ${side3}$. Find the perimeter of the triangle.",
            answer=perimeter,
            difficulty=(1200, 1300),
            topic="geometry/triangle_properties",
            solution=steps(
                f"The perimeter is the sum of all three sides",
                f"$P = {side1} + {side2} + {side3} = {perimeter}$"
            )
        )
    
    else:  # problem_type == 5
        # Exterior angle theorem (1400-1500)
        remote1 = randint(40, 70)
        remote2 = randint(40, 70)
        exterior = remote1 + remote2
        
        d = Diagram(width=350, height=250, padding=40)
        d.polygon([(0, 0), (8, 0), (3, 6)], labels=['A', 'B', 'C'])
        d.line((8, 0), (11, 0))
        d.text(0.5, 1, f'{remote1}°')
        d.text(2, 5, f'{remote2}°')
        d.text(9.5, 0.5, 'x')
        
        return problem(
            question=f"In triangle ABC, an exterior angle at vertex B is marked as $x$. The remote interior angles measure ${remote1}°$ and ${remote2}°$. Find the value of $x$ in degrees.",
            answer=exterior,
            difficulty=(1400, 1500),
            topic="geometry/triangle_properties",
            solution=steps(
                f"By the exterior angle theorem, an exterior angle equals the sum of the two remote interior angles",
                f"$x = {remote1}° + {remote2}°$",
                f"$x = {exterior}°$"
            ),
            image=d.render()
        )

emit(generate())

File: src/geometry_triangle_properties_medium.jl
# geometry - triangle_properties (medium)
# Generated: 2026-03-08T20:02:55.848360

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("geometry/triangle_properties")
    
    # Choose problem type randomly
    problem_type = choice([:angle_sum, :exterior_angle, :isosceles, :similar_triangles, :pythagorean_extended])
    
    if problem_type == :angle_sum
        # Basic angle sum with algebraic expression
        # Pick the three angles such that they sum to 180
        angle1_coeff = randint(2, 5)
        angle2_coeff = randint(2, 5)
        angle3_const = randint(20, 50)
        
        # We want angle1_coeff*x + angle2_coeff*x + angle3_const = 180
        # So x = (180 - angle3_const) / (angle1_coeff + angle2_coeff)
        ans = (180 - angle3_const) // (angle1_coeff + angle2_coeff)
        
        angle1 = angle1_coeff * ans
        angle2 = angle2_coeff * ans
        angle3 = angle3_const
        
        d = DiagramObj(300, 250)
        polygon!(d, [(50, 200), (250, 200), (150, 50)], fill="lightblue", stroke="black", stroke_width=2)
        point!(d, 50, 200, "A", :left)
        point!(d, 250, 200, "B", :right)
        point!(d, 150, 50, "C", :top)
        text!(d, 60, 180, "$(angle1_coeff)x", font_size=14)
        text!(d, 230, 180, "$(angle2_coeff)x", font_size=14)
        text!(d, 150, 70, "$(angle3_const)°", font_size=14)
        
        problem(
            question="In triangle ABC, angle A = $(angle1_coeff)x, angle B = $(angle2_coeff)x, and angle C = $(angle3_const)°. Find the value of x.",
            answer=ans,
            difficulty=(1200, 1400),
            solution=steps(
                "The sum of angles in a triangle equals 180°",
                sol("Equation", "$(angle1_coeff)x + $(angle2_coeff)x + $(angle3_const) = 180"),
                sol("Simplify", "$(angle1_coeff + angle2_coeff)x = $(180 - angle3_const)"),
                sol("Answer", "x = $(ans)")
            ),
            image=render(d),
            time=90
        )
        
    elseif problem_type == :exterior_angle
        # Exterior angle theorem with algebra
        remote1_coeff = randint(2, 4)
        remote2_const = randint(25, 55)
        exterior_coeff = randint(3, 6)
        
        # remote1_coeff*x + remote2_const = exterior_coeff*x
        # So (exterior_coeff - remote1_coeff)*x = remote2_const
        ans = remote2_const // (exterior_coeff - remote1_coeff)
        
        remote1 = remote1_coeff * ans
        remote2 = remote2_const
        exterior = exterior_coeff * ans
        adjacent = 180 - exterior
        
        d = DiagramObj(350, 250)
        polygon!(d, [(50, 200), (250, 200), (150, 50)], fill="lightyellow", stroke="black", stroke_width=2)
        line!(d, 250, 200, 320, 200, stroke="black", stroke_width=2, dash="5,5")
        point!(d, 50, 200, "A", :left)
        point!(d, 250, 200, "B", :bottom)
        point!(d, 150, 50, "C", :top)
        text!(d, 60, 180, "$(remote1_coeff)x", font_size=14)
        text!(d, 150, 70, "$(remote2_const)°", font_size=14)
        text!(d, 280, 210, "$(exterior_coeff)x", font_size=14)
        arc!(d, 250, 200, 20, 0, 25, stroke="blue", stroke_width=1.5)
        
        problem(
            question="In triangle ABC, the exterior angle at B measures $(exterior_coeff)x. The remote interior angles at A and C measure $(remote1_coeff)x and $(remote2_const)° respectively. Find x.",
            answer=ans,
            difficulty=(1300, 1500),
            solution=steps(
                "By the Exterior Angle Theorem, an exterior angle equals the sum of the two remote interior angles",
                sol("Equation", "$(exterior_coeff)x = $(remote1_coeff)x + $(remote2_const)"),
                sol("Simplify", "$(exterior_coeff - remote1_coeff)x = $(remote2_const)"),
                sol("Answer", "x = $(ans)")
            ),
            image=render(d),
            time=100
        )
        
    elseif problem_type == :isosceles
        # Isosceles triangle with perimeter constraint
        base = randint(8, 20)
        perimeter = randint(40, 80)
        
        # 2x + base = perimeter, so x = (perimeter - base)/2
        ans = (perimeter - base) // 2
        
        d = DiagramObj(300, 280)
        polygon!(d, [(100, 230), (200, 230), (150, 80)], fill="lightgreen", stroke="black", stroke_width=2)
        point!(d, 100, 230, "A", :left)
        point!(d, 200, 230, "B", :right)
        point!(d, 150, 80, "C", :top)
        text!(d, 150, 250, "$(base)", font_size=14)
        text!(d, 115, 150, "x", font_size=14)
        text!(d, 185, 150, "x", font_size=14)
        tick_marks!(d, 125, 155, 175, 155, count=1)
        tick_marks!(d, 175, 155, 200, 230, count=1)
        
        problem(
            question="Triangle ABC is isosceles with AC = BC = x. The base AB = $(base), and the perimeter is $(perimeter). Find x.",
            answer=ans,
            difficulty=(1100, 1300),
            solution=steps(
                "Perimeter = sum of all sides",
                sol("Equation", "x + x + $(base) = $(perimeter)"),
                sol("Simplify", "2x = $(perimeter - base)"),
                sol("Answer", "x = $(ans)")
            ),
            image=render(d),
            time=80
        )
        
    elseif problem_type == :similar_triangles
        # Similar triangles with proportions
        scale = randint(2, 4)
        small_side1 = randint(5, 12)
        small_side2 = randint(6, 15)
        
        large_side1 = scale * small_side1
        # We know one side of large triangle, find x (another side)
        ans = scale * small_side2
        
        d = DiagramObj(400, 280)
        # Small triangle
        polygon!(d, [(30, 240), (130, 240), (80, 180)], fill="none", stroke="blue", stroke_width=2)
        text!(d, 80, 260, "$(small_side1)", font_size=12, fill="blue")
        text!(d, 45, 210, "$(small_side2)", font_size=12, fill="blue")
        point!(d, 30, 240, "A", :left)
        point!(d, 130, 240, "B", :right)
        point!(d, 80, 180, "C", :top)
        
        # Large triangle
        polygon!(d, [(200, 240), (360, 240), (280, 120)], fill="none", stroke="red", stroke_width=2)
        text!(d, 280, 260, "$(large_side1)", font_size=12, fill="red")
        text!(d, 220, 180, "x", font_size=12, fill="red")
        point!(d, 200, 240, "D", :left)
        point!(d, 360, 240, "E", :right)
        point!(d, 280, 120, "F", :top)
        
        text!(d, 80, 30, "△ABC ~ △DEF", font_size=14)
        
        problem(
            question="Triangles ABC and DEF are similar. In △ABC, AB = $(small_side1) and AC = $(small_side2). In △DEF, DE = $(large_side1) and DF = x. Find x.",
            answer=ans,
            difficulty=(1400, 1600),
            solution=steps(
                "For similar triangles, corresponding sides are proportional",
                sol("Proportion", "\\frac{AB}{DE} = \\frac{AC}{DF}"),
                sol("Substitute", "\\frac{$(small_side1)}{$(large_side1)} = \\frac{$(small_side2)}{x}"),
                sol("Cross multiply", "$(small_side1) \\cdot x = $(large_side1) \\cdot $(small_side2)"),
                sol("Answer", "x = $(ans)")
            ),
            image=render(d),
            time=120
        )
        
    else  # pythagorean_extended
        # Right triangle with Pythagorean theorem and algebra
        leg1_coeff = randint(3, 8)
        leg2_coeff = randint(4, 9)
        hyp_const = randint(15, 35)
        
        # (leg1_coeff*x)^2 + (leg2_coeff*x)^2 = hyp_const^2
        # x^2 * (leg1_coeff^2 + leg2_coeff^2) = hyp_const^2
        sum_sq = leg1_coeff^2 + leg2_coeff^2
        ans_sq = hyp_const^2 // sum_sq
        ans = isqrt(ans_sq)
        
        # Only generate if perfect square
        if ans^2 == ans_sq && ans > 0
            leg1 = leg1_coeff * ans
            leg2 = leg2_coeff * ans
            
            d = DiagramObj(320, 280)
            polygon!(d, [(50, 230), (250, 230), (50, 80)], fill="lightcoral", stroke="black", stroke_width=2)
            right_angle!(d, 50, 230, 250, 230, 50, 80, size=15)
            point!(d, 50, 230, "A", :left)
            point!(d, 250, 230, "B", :right)
            point!(d, 50, 80, "C", :top)
            text!(d, 150, 250, "$(leg1_coeff)x", font_size=14)
            text!(d, 30, 155, "$(leg2_coeff)x", font_size=14)
            text!(d, 150, 140, "$(hyp_const)", font_size=14)
            
            problem(
                question="In right triangle ABC with right angle at A, AB = $(leg1_coeff)x, AC = $(leg2_coeff)x, and hypotenuse BC = $(hyp_const). Find x.",
                answer=ans,
                difficulty=(1500, 1700),
                solution=steps(
                    "Apply the Pythagorean theorem: a² + b² = c²",
                    sol("Equation", "($(leg1_coeff)x)^2 + ($(leg2_coeff)x)^2 = $(hyp_const)^2"),
                    sol("Expand", "$(leg1_coeff^2)x^2 + $(leg2_coeff^2)x^2 = $(hyp_const^2)"),
                    sol("Simplify", "$(sum_sq)x^2 = $(hyp_const^2)"),
                    sol("Solve", "x^2 = $(ans_sq)"),
                    sol("Answer", "x = $(ans)")
                ),
                image=render(d),
                time=150
            )
        else
            # Fallback to simpler angle sum problem
            angle1 = randint(35, 65)
            angle2 = randint(40, 70)
            ans = 180 - angle1 - angle2
            
            d = DiagramObj(300, 250)
            polygon!(d, [(50, 200), (250, 200), (150, 50)], fill="lightblue", stroke="black", stroke_width=2)
            point!(d, 50, 200, "A", :left)
            point!(d, 250, 200, "B", :right)
            point!(d, 150, 50, "C", :top)
            text!(d, 60, 180, "$(angle1)°", font_size=14)
            text!(d, 230, 180, "$(angle2)°", font_size=14)
            text!(d, 150, 70, "x", font_size=14)
            
            problem(
                question="In triangle ABC, angle A = $(angle1)°, angle B = $(angle2)°, and angle C = x. Find x.",
                answer=ans,
                difficulty=(1200, 1400),
                solution=steps(
                    "The sum of angles in a triangle equals 180°",
                    sol("Equation", "$(angle1) + $(angle2) + x = 180"),
                    sol("Simplify", "x = 180 - $(angle1) - $(angle2)"),
                    sol("Answer", "x = $(ans)")
                ),
                image=render(d),
                time=90
            )
        end
    end
end